### PR TITLE
Improve generation of "consumes" in ASP.NET Core, closes #420

### LINKED ISSUE #420: "Consumes" attribute on controller action ignored
"Consumes" attribute on controller action does not generate the swagger.json with a "path" with the indicated "consumes" property

### PR DESCRIPTION
**Pull request by RSuter**
Closes #420 

Tasks: 

- [x] Same for produces
- [x] If consumes/produces is empty then the consumes/produces entry can still be moved to global (?)

### CODE REVIEW

File: src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs [hunk 1]
```diff
@@ -146,14 +146,31 @@ public async Task<SwaggerDocument> GenerateAsync(object serviceProvider)
                     }
                 };
 
-                operations.Add(new Tuple<SwaggerOperationDescription, ApiDescription, MethodInfo>(operationDescription, apiDescription, method));
+                var consumes = apiDescription.SupportedRequestFormats
+                   .Select(f => f.MediaType).Concat(
+                       apiDescription.ActionDescriptor.ActionConstraints.OfType<ConsumesAttribute>().SelectMany(a => a.ContentTypes))
```
Comment (RSuter) [inline src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs]: @pranavkm Any idea why the ConsumesAttribute is not reported in SupportedRequestFormats by api explorer and I have to manually search for the attributes?

Comment (TheSilvermind) [reply src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs]: Very nice. Thank you very much.

Comment (RSuter) [reply src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs]: Pranav, thanks for clarifying, PR updated with custom input formatter

File: src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs [hunk 2]
```diff
@@ -145,14 +145,45 @@ public async Task<SwaggerDocument> GenerateAsync(object serviceProvider)
...
         }
 
-        private async Task<bool> AddOperationDescriptionsToDocumentAsync(SwaggerDocument document, Type controllerType, List<Tuple<SwaggerOperationDescription, ApiDescription, MethodInfo>> operations, SwaggerGenerator swaggerGenerator, SwaggerSchemaResolver schemaResolver)
+        private async Task<bool> AddOperationDescriptionsToDocumentAsync(SwaggerDocument document, Type controllerType,
+            List<Tuple<SwaggerOperationDescription, ApiDescription, MethodInfo, IEnumerable<string>, IEnumerable<string>>> operations,
+            SwaggerGenerator swaggerGenerator, SwaggerSchemaResolver schemaResolver)
         {
+            var globalConsumes = operations
+                .SelectMany(o => o.Item4)
+                .Where(c => operations.All(o => o.Item4.Count() == 0 || o.Item4.Contains(c)))
```
Comment (pranavkm) [inline src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs]: Super nit: Might be useful to start using `ValueTuple` so you could start doing `List<(SwaggerOperationDescription operationDescription, ApiDescription apiDescription, MethodInfo method, IEnumerable<string> consumes, IEnumerable<string> produces)>`. Makes it far more readable

File: src/NSwag.SwaggerGeneration.AspNetCore.Tests/Requests/ConsumesTests.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            var settings = new AspNetCoreToSwaggerGeneratorSettings();
+
+            // Act
+            var document = await GenerateDocumentAsync(settings, typeof(ConsumesController));
+            var json = document.ToJson();
+
+            // Assert
+            var operation = document.Operations.First(o => o.Operation.OperationId == "Consumes_ConsumesOnOperation").Operation;
+
+            Assert.DoesNotContain("foo/bar", document.Consumes);
```
Comment (pranavkm) [inline src/NSwag.SwaggerGeneration.AspNetCore.Tests/Requests/ConsumesTests.cs]: 👍

File: src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs (hunk 2 above)
Comment (RSuter) [reply src/NSwag.SwaggerGeneration.AspNetCore/AspNetCoreToSwaggerGenerator.cs]: Yes, I dont like this either - even using ValueTuple too much duplicated code for me :-) maybe an internal class is even better... But the bigger question is whether the new logic is correct :-). Thanks.

**Pull request merged**